contrib/babelfishpg_tsql/src/extendedproperty.c

```diff
@@ -0,0 +1,1139 @@
+				{
+					Form_pg_type	typeform;
+
+					tuple = SearchSysCache2(TYPENAMENSP,
+											CStringGetDatum(major_name),
+											Int16GetDatum(schema_id));
+					if (!HeapTupleIsValid(tuple))
+					{
+						ereport(ERROR,
+								(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+								 errmsg("Object is invalid. Extended properties are not permitted on '%s', or the object does not exist.", var_object_name)));
+					}
+
+					typeform = (Form_pg_type) GETSTRUCT(tuple);
+					owner_id = typeform->typowner;
+					ReleaseSysCache(tuple);
+
+					if (!has_all_permissions && owner_id != cur_user_id)
+					{
+						ereport(ERROR,
+								(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+								 errmsg("Cannot find the object \"%s\" because it does not exist or you do not have permissions.", var_object_name)));
+					}
+				}
+
+				if (!level2type)
+				{
+					type = asc_toupper(level1type, strlen(level1type));
+				}
+				else
+				{
+					char *temp;
+
+					if (strcmp(level1type, "table") != 0 &&
+						strcmp(level2type, "column") != 0)
+					{
+						ereport(ERROR,
+								(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+								 errmsg("Extended properties for object type %s are not currently supported in Babelfish.", level2type)));
+					}
+
+					var_object_name = psprintf("%s.%s.%s", level0name, level1name, level2name);
+					minor_name = level2name;
+
+					tuple = SearchSysCacheAttName(reloid, minor_name);
+					if (!HeapTupleIsValid(tuple))
+					{
+						ereport(ERROR,
+								(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+								 errmsg("Object is invalid. Extended properties are not permitted on '%s', or the object does not exist.", var_object_name)));
+					}
+					ReleaseSysCache(tuple);
+
+					temp = psprintf("%s %s", level1type, level2type);
+					type = asc_toupper(temp, strlen(temp));
+					pfree(temp);
+				}
+			}
+		}
+	}
+
+	/* insert/update/drop extended property */
+	rel = table_open(get_bbf_extended_properties_oid(), RowExclusiveLock);
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_extended_properties_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(db_id));
+	ScanKeyInit(&scanKey[1],
+				Anum_bbf_extended_properties_type,
+				BTEqualStrategyNumber, F_TEXTEQ,
+				CStringGetTextDatum(type));
+	ScanKeyInit(&scanKey[2],
+				Anum_bbf_extended_properties_schema_name,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyInit(&scanKey[3],
+				Anum_bbf_extended_properties_major_name,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(major_name));
+	ScanKeyInit(&scanKey[4],
+				Anum_bbf_extended_properties_minor_name,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(minor_name));
+	ScanKeyInit(&scanKey[5],
+				Anum_bbf_extended_properties_name,
+				BTEqualStrategyNumber, F_TEXTEQ,
+				CStringGetTextDatum(name));
+
+	scan = systable_beginscan(rel, get_bbf_extended_properties_idx_oid(), true,
+							  NULL, 6, scanKey);
+	if (HeapTupleIsValid(tuple = systable_getnext(scan)))
```

Comment:
I am not familiar with the coding pattern, do we need to check if it is returning only one row ?